File: Code/Day04.cs
using System;
using System.Collections.Generic;
using System.Linq;

namespace AoC.Code
{
    class Day04
    {
        public static int Solve(IEnumerable<IEnumerable<string>> indata)
        {
            return indata.Count(IsValid);
        }

        public static int Solve2(IEnumerable<IEnumerable<string>> indata)
        {
            return indata.Count(IsValid2);
        }


        public static bool IsValid(string phrase)
        {
            var words = phrase.Split(' ');
            return IsValid(words);
        }


        public static bool IsValid(IEnumerable<string> phrase)
        {
            return phrase.Distinct().Count() == phrase.Count();
        }


        public static bool IsValid2(string phrase)
        {
            var words = phrase.Split(' ');
            return IsValid2(words);
        }


        public static bool IsValid2(IEnumerable<string> phrase)
        {
            return IsValid(phrase) && NoAnagrams(phrase);
        }


        public static bool NoAnagrams(IEnumerable<string> phrase)
        {
            var charSorted = phrase.Select(word => new string(word.ToCharArray().OrderBy(c => c).ToArray()));
            return IsValid(charSorted);
        }
    }
}

File: AoC.FSharp/Day04.fs
namespace AoC.FSharp

module Day04 =
    let solve (s: string) = 
        let words = s.Split(' ') |> List.ofArray
        let d = List.distinct words
        d.Length = words.Length

    let solve2 (s: string) =
        let sortChars (s: string) =
            let chars = Seq.map(fun c -> c) s
            Seq.sort chars |> Seq.toArray |> System.String

        let words = s.Split(' ') |> List.ofArray
        let sortedWords = List.map (fun w -> sortChars w) words
        let distinct = Seq.distinct sortedWords |> List.ofSeq
        distinct.Length = sortedWords.Length